File: app/src/main/java/com/pluscubed/velociraptor/limit/FloatingView.java
package com.pluscubed.velociraptor.limit;

import android.animation.AnimatorSet;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.support.v4.view.animation.LinearOutSlowInInterpolator;
import android.support.v7.view.ContextThemeWrapper;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.WindowManager;
import android.widget.TextView;

import com.gigamole.library.ArcProgressStackView;
import com.pluscubed.velociraptor.R;
import com.pluscubed.velociraptor.utils.PrefUtils;
import com.pluscubed.velociraptor.utils.Utils;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FloatingView implements LimitView {

    @BindView(R.id.limit)
    View mLimitView;
    @Nullable
    @BindView(R.id.limit_label_text)
    TextView mLimitLabelText;
    @BindView(R.id.limit_text)
    TextView mLimitText;
    @BindView(R.id.speedometer)
    View mSpeedometerView;
    @BindView(R.id.arcview)
    ArcProgressStackView mArcView;
    @BindView(R.id.speed)
    TextView mSpeedometerText;
    @BindView(R.id.speedUnits)
    TextView mSpeedometerUnitsText;

    private LimitService mService;
    private WindowManager mWindowManager;
    private int mStyle;
    private View mFloatingView;
    private TextView mDebuggingText;

    public FloatingView(LimitService service) {
        this.mService = service;

        mWindowManager = (WindowManager) service.getSystemService(Context.WINDOW_SERVICE);
        inflateMonitor();
        inflateDebugging();

        updatePrefs();
    }

    @SuppressLint("InflateParams")
    private void inflateDebugging() {
        mDebuggingText = (TextView) LayoutInflater.from(new ContextThemeWrapper(mService, R.style.Theme_Velociraptor))
                .inflate(R.layout.floating_stats, null, false);

        WindowManager.LayoutParams debuggingParams = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                getWindowType(),
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                PixelFormat.TRANSLUCENT);

        debuggingParams.gravity = Gravity.BOTTOM;
        try {
            mWindowManager.addView(mDebuggingText, debuggingParams);
        } catch (Exception e) {
        }
    }

    private int getWindowType() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ?
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY :
                WindowManager.LayoutParams.TYPE_PHONE;
    }

    private void inflateMonitor() {
        int layout;
        switch (mStyle = PrefUtils.getSignStyle(mService)) {
            case PrefUtils.STYLE_US:
                layout = R.layout.floating_us;
                break;
            case PrefUtils.STYLE_INTERNATIONAL:
            default:
                layout = R.layout.floating_international;
                break;
        }

        mFloatingView = LayoutInflater.from(new ContextThemeWrapper(mService, R.style.Theme_Velociraptor))
                .inflate(layout, null, false);

        ButterKnife.bind(this, mFloatingView);

        WindowManager.LayoutParams params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                getWindowType(),
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT);
        params.gravity = Gravity.TOP | Gravity.START;
        params.alpha = PrefUtils.getOpacity(mService) / 100.0F;
        if (mWindowManager != null)
            try {
                mWindowManager.addView(mFloatingView, params);
            } catch (Exception e) {
            }
        mFloatingView.setOnTouchListener(new FloatingOnTouchListener());

        initMonitorPosition();

        final ArrayList<ArcProgressStackView.Model> models = new ArrayList<>();
        models.add(new ArcProgressStackView.Model("", 0,
                ContextCompat.getColor(mService, R.color.colorPrimary800),
                ContextCompat.getColor(mService, R.color.colorAccent)));
        mArcView.setTextColor(ContextCompat.getColor(mService, android.R.color.transparent));
        mArcView.setInterpolator(new FastOutSlowInInterpolator());
        mArcView.setModels(models);
    }


    private void initMonitorPosition() {
        if (mFloatingView == null) {
            return;
        }
        mFloatingView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                WindowManager.LayoutParams params = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();

                String[] split = PrefUtils.getFloatingLocation(mService).split(",");
                boolean left = Boolean.parseBoolean(split[0]);
                float yRatio = Float.parseFloat(split[1]);

                Point screenSize = new Point();
                mWindowManager.getDefaultDisplay().getSize(screenSize);
                params.x = left ? 0 : screenSize.x - mFloatingView.getWidth();
                params.y = (int) (yRatio * screenSize.y + 0.5f);

                try {
                    mWindowManager.updateViewLayout(mFloatingView, params);
                } catch (IllegalArgumentException ignore) {
                }

                mFloatingView.setVisibility(View.VISIBLE);

                mFloatingView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
            }
        });
    }

    @Override
    public void changeConfig() {
        initMonitorPosition();
    }

    @Override
    public void stop() {
        removeWindowView(mFloatingView);
        removeWindowView(mDebuggingText);
    }

    private void removeWindowView(View view) {
        if (view != null && mWindowManager != null)
            try {
                mWindowManager.removeView(view);
            } catch (IllegalArgumentException ignore) {
            }
    }

    private void animateViewToSideSlot() {
        Point screenSize = new Point();
        mWindowManager.getDefaultDisplay().getSize(screenSize);

        WindowManager.LayoutParams params = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();
        int endX;
        if (params.x + mFloatingView.getWidth() / 2 >= screenSize.x / 2) {
            endX = screenSize.x - mFloatingView.getWidth();
        } else {
            endX = 0;
        }

        PrefUtils.setFloatingLocation(mService, (float) params.y / screenSize.y, endX == 0);

        ValueAnimator valueAnimator = ValueAnimator.ofInt(params.x, endX)
                .setDuration(300);
        valueAnimator.setInterpolator(new LinearOutSlowInInterpolator());
        valueAnimator.addUpdateListener(animation -> {
            WindowManager.LayoutParams params1 = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();
            params1.x = (int) animation.getAnimatedValue();
            try {
                mWindowManager.updateViewLayout(mFloatingView, params1);
            } catch (IllegalArgumentException ignore) {
            }
        });

        valueAnimator.start();
    }

    @Override
    public void setSpeed(int speed, int percentOfWarning) {
        if (PrefUtils.getShowSpeedometer(mService) && mSpeedometerText != null) {
            mSpeedometerText.setText(Integer.toString(speed));
            mArcView.getModels().get(0).setProgress(percentOfWarning);
            mArcView.animateProgress();
        }
    }

    @Override
    public void setSpeeding(boolean speeding) {
        int colorRes = speeding ? R.color.red500 : R.color.primary_text_default_material_light;
        int color = ContextCompat.getColor(mService, colorRes);
        mSpeedometerText.setTextColor(color);
        mSpeedometerUnitsText.setTextColor(color);
    }

    @Override
    public void setDebuggingText(String text) {
        if (mDebuggingText != null && mDebuggingText.getVisibility() == View.VISIBLE) {
            mDebuggingText.setText(text);
        }
    }

    @Override
    public void setLimitText(String text) {
        if (mLimitText != null) {
            mLimitText.setText(text);
        }
    }

    @Override
    public void updatePrefs() {
        int prefStyle = PrefUtils.getSignStyle(mService);
        if (prefStyle != mStyle) {
            removeWindowView(mFloatingView);
            inflateMonitor();
        }
        mStyle = prefStyle;

        boolean debuggingEnabled = PrefUtils.isDebuggingEnabled(mService);
        if (mDebuggingText != null) {
            mDebuggingText.setVisibility(debuggingEnabled ? View.VISIBLE : View.GONE);
        }

        boolean speedometerShown = PrefUtils.getShowSpeedometer(mService);
        if (mSpeedometerView != null) {
            mSpeedometerView.setVisibility(speedometerShown ? View.VISIBLE : View.GONE);
        }

        boolean limitShown = PrefUtils.getShowLimits(mService);
        if (mLimitView != null) {
            mLimitView.setVisibility(limitShown ? View.VISIBLE : View.GONE);
        }

        if (mSpeedometerUnitsText != null) {
            mSpeedometerUnitsText.setText(Utils.getUnitText(mService));
        }

        if (mFloatingView != null) {
            WindowManager.LayoutParams layoutParams = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();
            layoutParams.alpha = PrefUtils.getOpacity(mService) / 100F;
            try {
                mWindowManager.updateViewLayout(mFloatingView, layoutParams);
            } catch (IllegalArgumentException ignore) {
            }
        }

        if (mLimitView != null && mLimitText != null) {
            float speedLimitSize = PrefUtils.getSpeedLimitSize(mService);
            float textSize = 0;
            float height = 0;
            float width = 0;

            switch (mStyle) {
                case PrefUtils.STYLE_US:
                    float cardSidePadding = (float) (2 + (1 - Math.cos(Math.toRadians(45))) * 2);
                    float cardTopBottomPadding = (float) (2 * 1.5 + (1 - Math.cos(Math.toRadians(45))) * 2);

                    width = speedLimitSize * 56;
                    height = speedLimitSize * 72;

                    width += cardSidePadding * 2;
                    height += cardTopBottomPadding * 2;

                    textSize = 28;

                    break;
                case PrefUtils.STYLE_INTERNATIONAL:
                    height = speedLimitSize * 64;
                    width = speedLimitSize * 64;

                    textSize = 0;
                    break;
            }

            ViewGroup.LayoutParams layoutParams = mLimitView.getLayoutParams();
            layoutParams.width = Utils.convertDpToPx(mService, width);
            layoutParams.height = Utils.convertDpToPx(mService, height);
            mLimitView.setLayoutParams(layoutParams);

            if (textSize != 0) {
                textSize *= speedLimitSize;
                mLimitText.setTextSize(TypedValue.COMPLEX_UNIT_SP, textSize);
            }

            float labelTextSize = 12 * speedLimitSize;
            if (mLimitLabelText != null) {
                mLimitLabelText.setTextSize(TypedValue.COMPLEX_UNIT_SP, labelTextSize);
            }
        }

        if (mSpeedometerView != null && mSpeedometerText != null && mSpeedometerUnitsText != null) {
            float speedometerSize = PrefUtils.getSpeedometerSize(mService);

            float width = 64 * speedometerSize;
            float height = 64 * speedometerSize;

            float textSize = 24 * speedometerSize;
            mSpeedometerText.setTextSize(TypedValue.COMPLEX_UNIT_SP, textSize);

            float labelTextSize = 12 * speedometerSize;
            mSpeedometerUnitsText.setTextSize(TypedValue.COMPLEX_UNIT_SP, labelTextSize);

            ViewGroup.LayoutParams layoutParams = mSpeedometerView.getLayoutParams();
            layoutParams.width = Utils.convertDpToPx(mService, width);
            layoutParams.height = Utils.convertDpToPx(mService, height);
            mSpeedometerView.setLayoutParams(layoutParams);
        }
    }

    @Override
    public void hideLimit(boolean hideLimit) {
        mLimitView.setVisibility(!PrefUtils.getShowLimits(mService) ? View.GONE : (hideLimit ? View.INVISIBLE : View.VISIBLE));
    }

    private class FloatingOnTouchListener implements View.OnTouchListener {

        private float mInitialTouchX;
        private float mInitialTouchY;
        private int mInitialX;
        private int mInitialY;
        private long mStartClickTime;
        private boolean mIsClick;

        private AnimatorSet fadeAnimator;
        private float initialAlpha;
        private ValueAnimator fadeOut;
        private ValueAnimator fadeIn;

        public FloatingOnTouchListener() {
            final WindowManager.LayoutParams params = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();
            fadeOut = ValueAnimator.ofFloat(params.alpha, 0.1F);
            fadeOut.setInterpolator(new FastOutSlowInInterpolator());
            fadeOut.setDuration(100);
            fadeOut.addUpdateListener(valueAnimator -> {
                params.alpha = (float) valueAnimator.getAnimatedValue();
                try {
                    mWindowManager.updateViewLayout(mFloatingView, params);
                } catch (IllegalArgumentException ignore) {
                }
            });
            fadeIn = fadeOut.clone();
            fadeIn.setFloatValues(0.1F, params.alpha);
            fadeIn.setStartDelay(5000);
        }

        @Override
        public boolean onTouch(View v, MotionEvent event) {
            final WindowManager.LayoutParams params = (WindowManager.LayoutParams) mFloatingView.getLayoutParams();

            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    mInitialTouchX = event.getRawX();
                    mInitialTouchY = event.getRawY();

                    mInitialX = params.x;
                    mInitialY = params.y;

                    mStartClickTime = System.currentTimeMillis();

                    mIsClick = true;
                    return true;
                case MotionEvent.ACTION_MOVE:
                    float dX = event.getRawX() - mInitialTouchX;
                    float dY = event.getRawY() - mInitialTouchY;
                    if ((mIsClick && (Math.abs(dX) > 10 || Math.abs(dY) > 10))
                            || System.currentTimeMillis() - mStartClickTime > ViewConfiguration.getLongPressTimeout()) {
                        mIsClick = false;
                    }

                    if (!mIsClick) {
                        params.x = (int) (dX + mInitialX);
                        params.y = (int) (dY + mInitialY);

                        try {
                            mWindowManager.updateViewLayout(mFloatingView, params);
                        } catch (IllegalArgumentException ignore) {
                        }
                    }
                    return true;
                case MotionEvent.ACTION_UP:
                    if (mIsClick && System.currentTimeMillis() - mStartClickTime <= ViewConfiguration.getLongPressTimeout()) {
                        if (fadeAnimator != null && fadeAnimator.isStarted()) {
                            fadeAnimator.cancel();
                            params.alpha = initialAlpha;
                            try {
                                mWindowManager.updateViewLayout(mFloatingView, params);
                            } catch (IllegalArgumentException ignore) {
                            }
                        } else {
                            initialAlpha = params.alpha;

                            fadeAnimator = new AnimatorSet();
                            fadeAnimator.play(fadeOut).before(fadeIn);
                            fadeAnimator.start();
                        }
                    } else {
                        animateViewToSideSlot();
                    }
                    return true;
            }
            return false;
        }
    }

}

File: app/src/main/java/com/pluscubed/velociraptor/limit/FloatingView.kt
package com.pluscubed.velociraptor.limit

import android.animation.AnimatorSet
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.PixelFormat
import android.graphics.Point
import android.os.Build
import android.util.TypedValue
import android.view.*
import android.widget.TextView
import androidx.appcompat.view.ContextThemeWrapper
import androidx.core.content.ContextCompat
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import androidx.interpolator.view.animation.LinearOutSlowInInterpolator
import butterknife.BindView
import butterknife.ButterKnife
import com.gigamole.library.ArcProgressStackView
import com.pluscubed.velociraptor.R
import com.pluscubed.velociraptor.utils.PrefUtils
import com.pluscubed.velociraptor.utils.Utils
import java.util.*

class FloatingView(private val service: LimitService) : LimitView {

    @BindView(R.id.limit)
    @JvmField
    var limitView: View? = null
    @BindView(R.id.limit_label_text)
    @JvmField
    var limitLabelText: TextView? = null
    @BindView(R.id.limit_text)
    @JvmField
    var limitText: TextView? = null
    @BindView(R.id.limit_source_text)
    @JvmField
    var limitSourceText: TextView? = null
    @BindView(R.id.speedometer)
    @JvmField
    var speedometerView: View? = null
    @BindView(R.id.arcview)
    @JvmField
    var arcView: ArcProgressStackView? = null
    @BindView(R.id.speed)
    @JvmField
    var speedometerText: TextView? = null
    @BindView(R.id.speedUnits)
    @JvmField
    var speedometerUnitsText: TextView? = null

    private val windowManager: WindowManager?
    private var style: Int = 0
    private var floatingView: View? = null
    private var debuggingText: TextView? = null

    private val windowType: Int
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        else
            WindowManager.LayoutParams.TYPE_PHONE

    init {

        windowManager = service.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        inflateMonitor()

        updatePrefs()
    }

    @SuppressLint("InflateParams")
    private fun inflateDebugging() {
        debuggingText =
            LayoutInflater.from(ContextThemeWrapper(service, R.style.Theme_Velociraptor))
                .inflate(R.layout.floating_stats, null, false) as TextView

        val debuggingParams = WindowManager.LayoutParams(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            windowType,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
            PixelFormat.TRANSLUCENT
        )

        debuggingParams.gravity = Gravity.BOTTOM
        try {
            windowManager!!.addView(debuggingText, debuggingParams)
        } catch (e: Exception) {
        }

    }

    private fun inflateMonitor() {
        style = PrefUtils.getSignStyle(service)
        val layout = when (style) {
            PrefUtils.STYLE_US -> R.layout.floating_region_us
            PrefUtils.STYLE_INTERNATIONAL -> R.layout.floating_region_international
            else -> R.layout.floating_region_international
        }

        // System.err errors here are printed by the ArcProgressStackView library
        floatingView = LayoutInflater.from(ContextThemeWrapper(service, R.style.Theme_Velociraptor))
            .inflate(layout, null, false)

        ButterKnife.bind(this, floatingView!!)

        val params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            windowType,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )
        params.gravity = Gravity.TOP or Gravity.START
        params.alpha = PrefUtils.getOpacity(service) / 100.0f
        if (windowManager != null)
            try {
                windowManager.addView(floatingView, params)
            } catch (e: Exception) {
            }

        floatingView!!.setOnTouchListener(FloatingOnTouchListener())

        initMonitorPosition()

        val models = ArrayList<ArcProgressStackView.Model>()
        models.add(
            ArcProgressStackView.Model(
                "", 0f,
                ContextCompat.getColor(service, R.color.colorPrimary800),
                ContextCompat.getColor(service, R.color.colorAccent)
            )
        )
        arcView!!.textColor = ContextCompat.getColor(service, android.R.color.transparent)
        arcView!!.interpolator = FastOutSlowInInterpolator()
        arcView!!.models = models
    }


    private fun initMonitorPosition() {
        if (floatingView == null) {
            return
        }
        floatingView!!.viewTreeObserver.addOnGlobalLayoutListener(object :
            ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                val params = floatingView!!.layoutParams as WindowManager.LayoutParams

                val split = PrefUtils.getFloatingLocation(service).split(",".toRegex())
                    .dropLastWhile { it.isEmpty() }.toTypedArray()
                val left = java.lang.Boolean.parseBoolean(split[0])
                val yRatio = java.lang.Float.parseFloat(split[1])

                val screenSize = Point()
                windowManager!!.defaultDisplay.getSize(screenSize)
                params.x = if (left) 0 else screenSize.x - floatingView!!.width
                params.y = (yRatio * screenSize.y + 0.5f).toInt()

                try {
                    windowManager.updateViewLayout(floatingView, params)
                } catch (ignore: IllegalArgumentException) {
                }

                floatingView!!.visibility = View.VISIBLE

                floatingView!!.viewTreeObserver.removeOnGlobalLayoutListener(this)
            }
        })
    }

    override fun changeConfig() {
        initMonitorPosition()
    }

    override fun stop() {
        removeWindowView(floatingView)
        removeWindowView(debuggingText)
    }

    private fun removeWindowView(view: View?) {
        if (view != null && windowManager != null)
            try {
                windowManager.removeView(view)
            } catch (ignore: IllegalArgumentException) {
            }

    }

    private fun animateViewToSideSlot() {
        val screenSize = Point()
        windowManager!!.defaultDisplay.getSize(screenSize)

        val params = floatingView!!.layoutParams as WindowManager.LayoutParams
        val endX: Int
        if (params.x + floatingView!!.width / 2 >= screenSize.x / 2) {
            endX = screenSize.x - floatingView!!.width
        } else {
            endX = 0
        }

        PrefUtils.setFloatingLocation(service, params.y.toFloat() / screenSize.y, endX == 0)

        val valueAnimator = ValueAnimator.ofInt(params.x, endX)
            .setDuration(300)
        valueAnimator.interpolator = LinearOutSlowInInterpolator()
        valueAnimator.addUpdateListener { animation ->
            val params1 = floatingView!!.layoutParams as WindowManager.LayoutParams
            params1.x = animation.animatedValue as Int
            try {
                windowManager.updateViewLayout(floatingView, params1)
            } catch (ignore: IllegalArgumentException) {
            }
        }

        valueAnimator.start()
    }

    override fun setSpeed(speed: Int, percentOfWarning: Int) {
        if (PrefUtils.getShowSpeedometer(service) && speedometerText != null) {
            speedometerText!!.text = String.format("%d", speed)
            arcView!!.models[0].progress = percentOfWarning.toFloat()
            arcView!!.animateProgress()
        }
    }

    override fun setSpeeding(speeding: Boolean) {
        val redColor = ContextCompat.getColor(service, R.color.red500)

        val typedValue = TypedValue()
        speedometerText?.context?.theme?.resolveAttribute(R.attr.colorOnBackground, typedValue, true)
        val textColor = speedometerText?.context?.let { ContextCompat.getColor(it, typedValue.resourceId) }
                ?: Color.BLACK;

        val color = if (speeding) redColor else textColor
        speedometerText?.setTextColor(color)
        speedometerUnitsText?.setTextColor(color)
    }

    override fun setDebuggingText(text: String) {
        if (debuggingText != null) {
            debuggingText?.text = text
        }
    }

    override fun setLimit(limit: String, source: String) {
        limitText?.text = limit
        limitSourceText?.text = source
    }

    override fun updatePrefs() {
        val prefStyle = PrefUtils.getSignStyle(service)
        if (prefStyle != style) {
            removeWindowView(floatingView)
            inflateMonitor()
        }
        style = prefStyle

        val debuggingEnabled = PrefUtils.isDebuggingEnabled(service)
        if (debuggingEnabled && debuggingText == null) {
            inflateDebugging()
        } else if (!debuggingEnabled && debuggingText != null) {
            removeWindowView(debuggingText)
            debuggingText = null
        }

        val speedometerShown = PrefUtils.getShowSpeedometer(service)
        speedometerView?.visibility = if (speedometerShown) View.VISIBLE else View.GONE

        val limitShown = PrefUtils.getShowLimits(service)
        limitView?.visibility = if (limitShown) View.VISIBLE else View.GONE

        speedometerUnitsText?.text = Utils.getUnitText(service)

        if (floatingView != null) {
            val layoutParams = floatingView!!.layoutParams as WindowManager.LayoutParams
            layoutParams.alpha = PrefUtils.getOpacity(service) / 100f
            try {
                windowManager!!.updateViewLayout(floatingView, layoutParams)
            } catch (ignore: IllegalArgumentException) {
            }

        }

        // TODO: Remove hard-coded base text sizes
        if (limitView != null && limitText != null) {
            val speedLimitSize = PrefUtils.getSpeedLimitSize(service)
            var textSize = 0f
            var height = 0f
            var width = 0f

            when (style) {
                PrefUtils.STYLE_US -> {
                    val cardSidePadding = (2 + (1 - Math.cos(Math.toRadians(45.0))) * 2).toFloat()
                    val cardTopBottomPadding =
                        (2 * 1.5 + (1 - Math.cos(Math.toRadians(45.0))) * 2).toFloat()

                    width = speedLimitSize * 56
                    height = speedLimitSize * 80

                    width += cardSidePadding * 2
                    height += cardTopBottomPadding * 2

                    textSize = 28f
                }
                PrefUtils.STYLE_INTERNATIONAL -> {
                    height = speedLimitSize * 64
                    width = speedLimitSize * 64

                    textSize = 0f
                }
            }

            val layoutParams = limitView!!.layoutParams
            layoutParams.width = Utils.convertDpToPx(service, width)
            layoutParams.height = Utils.convertDpToPx(service, height)
            limitView!!.layoutParams = layoutParams

            if (textSize != 0f) {
                textSize *= speedLimitSize
                limitText!!.setTextSize(TypedValue.COMPLEX_UNIT_DIP, textSize)
            }

            val labelTextSize = 12 * speedLimitSize
            limitLabelText?.setTextSize(TypedValue.COMPLEX_UNIT_DIP, labelTextSize)

            val sourceTextSize = 8 * speedLimitSize
            limitSourceText?.setTextSize(TypedValue.COMPLEX_UNIT_DIP, sourceTextSize)
        }

        if (speedometerView != null && speedometerText != null && speedometerUnitsText != null) {
            val speedometerSize = PrefUtils.getSpeedometerSize(service)

            val width = 64 * speedometerSize
            val height = 64 * speedometerSize

            val textSize = 24 * speedometerSize
            speedometerText!!.setTextSize(TypedValue.COMPLEX_UNIT_DIP, textSize)

            val labelTextSize = 12 * speedometerSize
            speedometerUnitsText!!.setTextSize(TypedValue.COMPLEX_UNIT_DIP, labelTextSize)

            val layoutParams = speedometerView!!.layoutParams
            layoutParams.width = Utils.convertDpToPx(service, width)
            layoutParams.height = Utils.convertDpToPx(service, height)
            speedometerView!!.layoutParams = layoutParams
        }
    }

    override fun hideLimit(hideLimit: Boolean) {
        limitView!!.visibility =
            if (!PrefUtils.getShowLimits(service)) View.GONE else if (hideLimit) View.INVISIBLE else View.VISIBLE
    }

    private inner class FloatingOnTouchListener : View.OnTouchListener {

        private var mInitialTouchX: Float = 0.toFloat()
        private var mInitialTouchY: Float = 0.toFloat()
        private var mInitialX: Int = 0
        private var mInitialY: Int = 0
        private var mStartClickTime: Long = 0
        private var mIsClick: Boolean = false

        private var fadeAnimator: AnimatorSet? = null
        private var initialAlpha: Float = 0.toFloat()
        private val fadeOut: ValueAnimator
        private val fadeIn: ValueAnimator

        init {
            val params = floatingView!!.layoutParams as WindowManager.LayoutParams
            fadeOut = ValueAnimator.ofFloat(params.alpha, 0.1f)
            fadeOut.interpolator = FastOutSlowInInterpolator()
            fadeOut.duration = 100
            fadeOut.addUpdateListener { valueAnimator ->
                params.alpha = valueAnimator.animatedValue as Float
                try {
                    windowManager!!.updateViewLayout(floatingView, params)
                } catch (ignore: IllegalArgumentException) {
                }
            }
            fadeIn = fadeOut.clone()
            fadeIn.setFloatValues(0.1f, params.alpha)
            fadeIn.startDelay = 5000
        }

        override fun onTouch(v: View, event: MotionEvent): Boolean {
            val params = floatingView!!.layoutParams as WindowManager.LayoutParams

            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    mInitialTouchX = event.rawX
                    mInitialTouchY = event.rawY

                    mInitialX = params.x
                    mInitialY = params.y

                    mStartClickTime = System.currentTimeMillis()

                    mIsClick = true
                    return true
                }
                MotionEvent.ACTION_MOVE -> {
                    val dX = event.rawX - mInitialTouchX
                    val dY = event.rawY - mInitialTouchY
                    if (mIsClick && (Math.abs(dX) > 10 || Math.abs(dY) > 10) || System.currentTimeMillis() - mStartClickTime > ViewConfiguration.getLongPressTimeout()) {
                        mIsClick = false
                    }

                    if (!mIsClick) {
                        params.x = (dX + mInitialX).toInt()
                        params.y = (dY + mInitialY).toInt()

                        try {
                            windowManager!!.updateViewLayout(floatingView, params)
                        } catch (ignore: IllegalArgumentException) {
                        }

                    }
                    return true
                }
                MotionEvent.ACTION_UP -> {
                    if (mIsClick && System.currentTimeMillis() - mStartClickTime <= ViewConfiguration.getLongPressTimeout()) {
                        if (fadeAnimator != null && fadeAnimator!!.isStarted) {
                            fadeAnimator!!.cancel()
                            params.alpha = initialAlpha
                            try {
                                windowManager!!.updateViewLayout(floatingView, params)
                            } catch (ignore: IllegalArgumentException) {
                            }

                        } else {
                            initialAlpha = params.alpha

                            fadeAnimator = AnimatorSet()
                            fadeAnimator!!.play(fadeOut).before(fadeIn)
                            fadeAnimator!!.start()
                        }
                    } else {
                        animateViewToSideSlot()
                    }
                    return true
                }
            }
            return false
        }
    }

}
